Use retry with split in GpuCachedDoublePassWindowIterator

andygrove opened the pull request:
Closes https://github.com/NVIDIA/spark-rapids/issues/8388

This PR updates GpuCachedDoublePassWindowIterator to use retry with splits for pre and post-processing.

andygrove commented:
build

andygrove commented:
build

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:
```diff
@@ -1661,6 +1661,8 @@ class GpuCachedDoublePassWindowIterator(
 
   override def isRunningBatched: Boolean = true
 
+  private var firstPassIter: Option[Iterator[(Array[cudf.ColumnVector], ColumnarBatch)]] = None
```
Could we get an explanation of what is happening with this iterator?.  Why is there a pair of what appears to be two tables of data?

revans2 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:
```diff
@@ -1694,6 +1696,12 @@ class GpuCachedDoublePassWindowIterator(
       waitingForFirstPass.foreach(_.close())
       waitingForFirstPass = None
+
+      firstPassIter.foreach(_.foreach(_._1.foreach(_.close())))
```
Why are we pulling off of the results from the iterators to close them?  This is not common unless the Iterators are backed by a cache. But then why have it be an Iterator instead of a List or ArrayBuffer or something like that.

Are we risking leaking data if the retry iterator is not fully drained?  If so then there are a number of other bugs that we need to fix.

revans2 replied:
Okay I filed https://github.com/NVIDIA/spark-rapids/issues/8482 for this. Could you add a comment here about it so that we can find a better way to clean this up?

revans2 reviewed (commented):
Sorry I didn't finish the review.  Will look at it more soon once I fix the other places that we are trying to use withRetry in a way that is not going to close the Iterator that is returned.

revans2 reviewed (dismissed):
I think this code is okay, but I think we could make it much cleaner if we had a better way to close the iterators and divide things up into sub-iterators.  Perhaps it would be better to wait until 23.08 and we have fixed the leak that you are worried about.

andygrove commented:
> I think this code is okay, but I think we could make it much cleaner if we had a better way to close the iterators and divide things up into sub-iterators. Perhaps it would be better to wait until 23.08 and we have fixed the leak that you are worried about.

Sounds good. I have moved this to draft for now.

andygrove commented:
build
